sysl2/sysl/validate.go

```diff
@@ -0,0 +1,145 @@
+package main
+
+import (
+	"github.com/anz-bank/sysl/src/proto"
+	"github.com/sirupsen/logrus"
+	"strings"
+)
+
+type content struct {
+	name     string
+	typeRef  string
+	children map[string]content
+}
+
+type View struct {
+	name       string
+	body       map[string]content
+	returnType string
+	rawData    *sysl.View
+}
+type Views map[string]View
+
+func typeRef(expr *sysl.Expr) string {
+	if expr.GetLiteral() != nil {
+		return "string"
+	}
+	if expr.GetList() != nil {
+		return "list"
+	}
+	if call := expr.GetCall(); call != nil {
+		if _, exists := GoFuncMap[call.GetFunc()]; exists {
+			return "string"
+		}
+		return "Transform(Unhandled)"
+	}
+	if varName := expr.GetName(); varName != "" {
+		return varName
+	}
+	if expr.GetBinexpr() != nil {
+		return "BinExpr"
+	}
+	if expr.GetIfelse() != nil {
+		return "Conditional"
+	}
+	return "Yet unknown"
+}
+
+func readView(stmt *sysl.Expr_Transform_Stmt) content {
+	if tfm := stmt.GetAssign().GetExpr().GetTransform(); tfm != nil {
+		children := map[string]content{}
+
+		for _, tfmStmt := range tfm.GetStmt() {
+			if tfmStmt.GetAssign() != nil {
+				child := readView(tfmStmt)
+				children[child.name] = child
+			}
+		}
+
+		return content{
+			name:     stmt.GetAssign().GetName(),
+			typeRef:  typeRef(stmt.GetAssign().GetExpr()),
+			children: children,
+		}
+	} else {
+		return content{
+			name:    stmt.GetAssign().GetName(),
+			typeRef: typeRef(stmt.GetAssign().GetExpr()),
+		}
+	}
+}
+
+func returnType(syslType *sysl.Type) string {
```

Comment:
Returntype is probably not the right name for it

Reply:
Changed